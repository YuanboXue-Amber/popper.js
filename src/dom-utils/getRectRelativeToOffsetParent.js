// @flow
import type { Rect, VirtualElement } from '../types';
import getBoundingClientRect from './getBoundingClientRect';
import listScrollParents from './listScrollParents';
import getScrollSum from './getScrollSum';
import getOffsetParent from './getOffsetParent';
import unwrapVirtualElement from './unwrapVirtualElement';
import getWindow from './getWindow';

// Returns the width, height and offsets of the provided element relative to the
// offsetParent
export default (
  element: Element | VirtualElement,
  isFixed: boolean = false
): Rect => {
  const unwrappedElement = unwrapVirtualElement(element);
  const win = getWindow(unwrappedElement);
  const rect = getBoundingClientRect(element);
  const scrollParentsScrollSum = getScrollSum(
    listScrollParents(unwrappedElement)
  );
  const directOffsetParent = getOffsetParent(unwrappedElement);
  const directOffsetParentRect =
    directOffsetParent instanceof win.Element && !isFixed
      ? getBoundingClientRect(directOffsetParent)
      : { left: 0, top: 0 };

  const ancestorOffsetParents = isFixed ? [directOffsetParent] : [];
  let currentOffsetParent = directOffsetParent;

  while (currentOffsetParent instanceof win.Element) {
    currentOffsetParent = getOffsetParent(currentOffsetParent);
    ancestorOffsetParents.push(currentOffsetParent);
  }

  const ancestorOffsetParentScrollSum = getScrollSum(ancestorOffsetParents);

  const width = rect.width;
  const height = rect.height;
  const x =
    rect.left +
    scrollParentsScrollSum.scrollLeft -
    ancestorOffsetParentScrollSum.scrollLeft -
    directOffsetParentRect.left;
  const y =
    rect.top +
    scrollParentsScrollSum.scrollTop -
    ancestorOffsetParentScrollSum.scrollTop -
    directOffsetParentRect.top;

  return { width, height, x, y };
};
