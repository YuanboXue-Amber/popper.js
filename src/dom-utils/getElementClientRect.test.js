// @flow
import getRectRelativeToOffsetParent from './getRectRelativeToOffsetParent';

it('returns an object containing x, y, width, and height properties', () => {
  const element = document.createElement('div');
  expect(getRectRelativeToOffsetParent(element)).toMatchSnapshot();
});

it('works with margins', () => {
  const element = document.createElement('div');
  Object.assign(element.style, { margin: '10px' });
  expect(getRectRelativeToOffsetParent(element)).toMatchSnapshot();
});
