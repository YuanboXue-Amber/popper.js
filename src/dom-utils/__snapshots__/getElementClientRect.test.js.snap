// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns an object containing x, y, width, and height properties 1`] = `
Object {
  "height": 0,
  "width": 0,
  "x": 0,
  "y": 0,
}
`;

exports[`works with margins 1`] = `
Object {
  "height": 20,
  "width": 20,
  "x": -10,
  "y": -10,
}
`;
